### ISSUE TITLE
Test OSX Gitian descriptor

### ISSUE THREAD

**Issue by JeremyRand** [namecoin/namecoin-core#38]
As far as I can tell, no one's tested the OSX Gitian descriptor.  It needs the remote URL changed.  Not sure what else.

**Comment by JeremyRand** [comment]
Does anyone have opinions on whether we want this to block nc0.12.0?  moa posted instructions for doing a depends build on OS X a while back.  While those instructions don't result in reproducible hashes, 0.3.80 isn't reproducible either, so it's not a regression.  If, for whatever reason, we need to hurry in order to deal with the scheduled hardfork safely, I wouldn't object to deferring this task, and just having @cassiniNMC do a depends build for OS X users (assuming moa's instructions work).

Opinions?

**Comment by cassiniNMC** [comment]
Unfortunately moa didn't explain where to get the latest libraries from, let alone how to build a GUI app. For the moment being I therefore prefer [Bitcoin's manual build process](https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md) (ignoring the paragraph that states you need XCode. This is simply incorrect). For the GUI app I'd stick to maxpower's method.

**Comment by JeremyRand** [comment]
@cassiniNMC Which libraries are needed, which aren't downloaded by the depends build?

**Comment by JeremyRand** [comment]
Okay, additional issue.  OS X builds require Xcode inputs from Apple's website.  It seems that Apple's developer downloads website blacklists Tor Browser users, the downloads don't use TLS, and no hashes or signatures are provided.  Until Apple unblocks Tor Browser users, and either enables TLS or provides hashes, I have no interest in trying to create Gitian OS X binaries.

**Comment by cassiniNMC** [comment]
> Which libraries are needed,  

For building namecoind: see https://wiki.namecoin.org/index.php?title=Build_Namecoin_From_Source#OS_X
In addition to that you need qt5 for building brandonrobertz' GUI PR https://github.com/namecoin/namecoin-core/pull/67 , and Boost and berkeley-db4 need to be an older version there.


> which aren't downloaded by the depends build?  

As [you already complained](https://forum.namecoin.info/viewtopic.php?p=15673#p15673) in moa's thread he didn't explain at all how to download the libraries.
What exactly is it moa calls a "depends build"? This is a very vague expression particularly in OSX which has three levels of "deployability" of an executable.


> OS X builds require Xcode inputs from Apple's website.  

No, they don't. XCode can actually interfere with Homebrew or with the Command Line Developer Tools. You do need the latest Command Line Developer Tools, though. There are at least two methods for downloading those, one of them is using a command with the misleading name "xcode-select". This has nothing to do with XCode in this case. tl;dr: see https://wiki.namecoin.org/index.php?title=Build_Namecoin_From_Source#OS_X , step 1

**Comment by JeremyRand** [comment]
> For building namecoind: see https://wiki.namecoin.org/index.php?title=Build_Namecoin_From_Source#OS_X

Most of the dependencies listed there are libraries handled by the depends build.  (At least, they are on Linux; I don't know why OS X would be different.)

> In addition to that you need qt5 for building brandonrobertz' GUI PR #67

Qt is needed regardless of the name management tab; the only reason to exclude it if if you don't want Namecoin-Qt to be built.

>  Boost and berkeley-db4 need to be an older version there.

The depends build uses the older versions automatically.

> As you already complained in moa's thread he didn't explain at all how to download the libraries.

My comment that you linked was about build tools (e.g. autoconf); it had nothing to do with downloading the libraries.  moa did provide instructions on downloading the libraries; it's step 4 of his instructions.

> What exactly is it moa calls a "depends build"?

A depends build downloads and builds the libraries from source and statically links them, rather than using the OS's library packages with dynamic linking.  Depends builds are important for consensus-critical code, and are also needed for reproducible builds (unless the library OS packages are themselves reproducible, which is what Debian is aiming for).

> No, they don't. XCode can actually interfere with Homebrew or with the Command Line Developer Tools. You do need the latest Command Line Developer Tools, though. There are at least two methods for downloading those, one of them is using a command with the misleading name "xcode-select". This has nothing to do with XCode in this case. (I must admit I have no idea how to obtain those Command Line Developer Tools within a Gitian build). Also see https://wiki.namecoin.org/index.php?title=Build_Namecoin_From_Source#OS_X , step 1

I don't see how Xcode can interfere with the command-line developer tools, given that Xcode isn't even installed on the Gitian guest.  The only reason Xcode is needed is to extract the OS X SDK, which is then placed in a .tar and used as an input to Gitian.  If there's a way to extract that SDK from the command-line developer tools, it's not mentioned in any of the Bitcoin build documentation.

**Comment by JeremyRand** [comment]
Looks like there may be issues with depends builds, see https://github.com/namecoin/namecoin-core/issues/83#issuecomment-204255326 .  Is anyone else either successfully doing depends builds of Namecoin Core on OS X, or having similar issues?

**Comment by JeremyRand** [comment]
Replying to myself:

> Okay, additional issue. OS X builds require Xcode inputs from Apple's website. It seems that Apple's developer downloads website blacklists Tor Browser users, the downloads don't use TLS, and no hashes or signatures are provided. Until Apple unblocks Tor Browser users, and either enables TLS or provides hashes, I have no interest in trying to create Gitian OS X binaries.

Upon further consideration, the lack of TLS, hashes, and signatures isn't quite as much a big deal as I initially thought, because if a subset of our Gitian builders received a tampered-with Xcode binary, then the build would fail to reproduce for everyone, and we would know something was wrong.  So the only way that this would result in an undetected security issue is if *all* of our builders got the same tampered-with Xcode binary.  It's definitely worth a big scary warning on the downloads page, but I don't think it precludes distributing binaries at all.

The blacklisting of Tor Browser is still a dealbreaker for me, though.  It's likely that the Tor Browser issue is a result of https://trac.torproject.org/projects/tor/ticket/15954 .  So, not necessarily Apple's fault.

**Comment by JeremyRand** [comment]
@cassiniNMC Do I recall correctly that the public 0.3.80 binaries are from you?  Are you currently capable of building a non-reproducible OS X binary of the Namecoin Core Qt GUI?  If so, and if you can document the build instructions, then I think this isn't a regression from 0.3.80 (which wasn't reproducible), so we could defer this until the nc0.12.2 milestone (after BIP9 is dealt with).  Opinions?

**Comment by JeremyRand** [comment]
Pinging @cassiniNMC again.

@jonasbits and @josephbisch , I understand that you've made some progress on this front recently?

**Comment by cassiniNMC** [comment]
(a)
> Do I recall correctly that the public 0.3.80 binaries are from you?

(b)
> Are you currently capable of building a non-reproducible OS X binary of the Namecoin Core Qt GUI?

@JeremyRand These two questions address two different kettles of fish:
(a) This is a fully deployed OSX App as produced by the standard *macdeployqt* tool (or maxpower's infamous script). Furthermore, Homebrew needs to be in compatibility mode (i.e. using --build-bottle) while downloading and compiling the libraries, otherwise the executable raises a SIGILL exception on some Intel CPUs. I haven't tried all this on Mac OS 10.11.x El Capitan yet, therefore I'm not absolutely sure if this method still works without an Apple Developer Licence. (And yes, it was me who provided the publicly available 0.3.80 Qt version.)
(b) addresses the easier variant. Homebrew can be run in standard mode, and deployment is a fairly simple task. The resulting executable runs only on the build machine, of course.

As of Tuesday (May 3rd) I'll have more time to spare. I'm going to try version (b) first. If this works I'll be able to update the wiki with these new GUI build instructions.

**Comment by JeremyRand** [comment]
FYI @jonasbits is reporting that he's gotten Gitian builds for OS X to succeed for Namecoin Core.  See https://github.com/namecoin/namecoin-core/pull/67#issuecomment-223847706 .

**Comment by JeremyRand** [comment]
@jonasbits Can you remove the change to the repo URL in https://github.com/namecoin/namecoin-core/commit/1069cccdada095f3891d26d83759514b5b8654db and then submit it as a PR?  I haven't tested it (haven't had a chance to use an OS X machine on the campus), but the changes look straightforward.

**Comment by JeremyRand** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

The OSX Gitian descriptor should work now thanks to #153.  Closing
this.  If any additional problems come up, please create a new issue.

-----BEGIN PGP SIGNATURE-----

iQIcBAEBCgAGBQJZBB7nAAoJELPy0WV4bWVw46EP/j7YecZjwBJ3zSQTP8g/gTu7
adpSyBZE5VvOGfUg+WsUUORLTRDRUwRXxw05879hwvCf9PcMgRn84me86JELHiS5
yobO0cWtcxgVs+UWPZwiX8fZkngaQFbbd8D6ypnvDMNmndBsFMpQwCFlfctlinIi
28tn6bmFLMPdRGmoaGZHR9hR7lyVAGXqw1BWZqzSQM8stBI62cRMtpKQXAvQau8j
EkN0gw0G6QuBo/mF0g0lCTMZunN8xnz2pppOj6Nk7AIZzv60T1Y5Zx3PQwo9sY6N
ThOr7e+q7fh92eMXcrIiLPEqjsJU6g9jB89byxwUhkSe4TvtfMjM1LnDTzmCmNKI
/4dr2jB6tY20HVLxCYFwIfIjrfDuyIknt3KMn4Z3lBILkHp0YWyK2t76nPeOQcqT
oAJIVE0lZlgEWe4SxFz63Ls/J01P5GmvxFic7pOddsD6HNVMnRF4RMKs2fafdqRv
XpJqT/CdZQhY4pC4aijUyVcBM5e+vJNQGv7gUGXb6Nq1ew/Dzp+QZWKBLWWU/PHQ
5QX9rCqAof+7tpUbIv2JbJjQjfK8o0qWo8EczMUx3fzGOOSbysRqMdCTDnY4H6Aa
rg9rsyqYPWaYanuNTwuquJz3K5L363z37NJNGcxhPOsAmerz/zh5wqJeZ1LjuScq
KKpVEo90hB9SW9C/NjGs
=KpPO
-----END PGP SIGNATURE-----
